randovania/generator/old_generator_reach.py

```diff
@@ -222,52 +222,94 @@ def _calculate_safe_nodes(self) -> None:
         for component in self._digraph.strongly_connected_components():
             if self._state.node.node_index in component:
                 assert self._safe_nodes is None
-                self._safe_nodes = _SafeNodes(sorted(component), component)
+                self._safe_nodes = _SafeNodes(
+                    sorted(component),
+                    set(component),
+                )
 
         assert self._safe_nodes is not None
 
     def _calculate_reachable_paths(self) -> None:
         if self._reachable_paths is not None:
             return
 
-        all_nodes = self.all_nodes
         context = self.node_context()
 
-        @functools.cache
-        def _is_collected(target: int) -> int:
-            node = all_nodes[target]
-            if node.is_resource_node():
-                assert isinstance(node, GraphOrResourceNode)
-                if node.is_collected(context):
-                    return 0
+        if isinstance(self._game, WorldGraph):
+            graph_nodes = self._game.nodes
+
+            @functools.cache
+            def _is_collected(target: int) -> int:
+                return not graph_nodes[target].is_collected(context)
+        else:
+            db_nodes = typing.cast("tuple[Node, ...]", self._game.region_list.all_nodes)
+
+            @functools.cache
+            def _is_collected(target: int) -> int:
+                node = db_nodes[target]
+                if node.is_resource_node():
+                    if typing.cast("ResourceNode", node).is_collected(context):
+                        return 0
+                    else:
+                        return 1
                 else:
-                    return 1
-            else:
-                return 0
+                    return 0
+
+        self._reachable_is_collected = _is_collected
 
         def weight(source: int, target: int, attributes: graph_module.GraphData) -> int:
             return _is_collected(target)
 
-        self._reachable_costs, self._reachable_paths = self._digraph.multi_source_dijkstra(
-            {self._state.node.node_index},
-            weight=weight,
-        )
+        if self._digraph.has_multi_source_dijkstra:
+            costs, self._reachable_paths = self._digraph.multi_source_dijkstra(
+                {self._state.node.node_index},
+                weight=weight,
+            )
+            self._reachable_costs = typing.cast("dict[int, int | None]", costs)
+        else:
+            # Having to calculate the costs in python is slower,
```

Comment:
does rustworkx not support calculating the costs?

Reply:
> A dictionary of the shortest paths from the provided node where the key is the node index of the end of the path and the **value is the cost/sum of the weights of path**

?

Reply:
We're calling the one they returns a dict of index to list of indices, indicating the path. I know we use both info, but can't remember on top of my head if the path is essential